modules/service/src/main/scala/lucuma/odb/service/TimeAccountingService.scala

```diff
@@ -3,59 +3,328 @@
+
+        private val computeInvoice: F[TimeCharge.Invoice] = {
+          def invoice(corrections: List[Correction]): StateT[F, TimeAccountingState, TimeCharge.Invoice] =
+            for {
+              ini <- StateT.get[F, TimeAccountingState]
+              day <- daylightDiscounts
+              qa  <- qaDiscounts
+              // ... other discounts here ...
+              fin <- StateT.get[F, TimeAccountingState]
+            } yield TimeCharge.Invoice(ini.charge, day ++ qa, fin.charge.corrected(corrections))
 
+          for {
+            s <- initialState
+            c <- session.execute(SelectCorrection)(visitId)
+            i <- invoice(c).runA(s)
+          } yield i
+        }
+
+        private def updateInvoice(inv: TimeCharge.Invoice): F[Unit] =
+          for {
+            _ <- updateDiscounts(inv.discounts)
+            _ <- session.execute(UpdateTimeAccounting)((visitId, inv.executionTime, inv.finalCharge))
+          } yield ()
+
+        // Delete existing discounts (if any) and write the new ones.
+        private def updateDiscounts(discounts: List[TimeCharge.DiscountEntry]): F[Unit] =
+          for {
+            _ <- session.execute(DeleteDiscountEntries)(visitId)
+            _ <- discounts.traverse_(storeDiscount)
+          } yield ()
+
+        // Stores the discount entry.
+        private def storeDiscount(d: TimeCharge.DiscountEntry): F[Unit] =
+          for {
+            id <- session.unique(StoreDiscountEntry)(visitId, d)
+            _  <- storeDiscountDetail(id)(d)
+          } yield ()
+
+        // Stores any information specific to particular discount entry types.
+        private def storeDiscountDetail(id: Long): TimeCharge.DiscountEntry => F[Unit] = {
+          case TimeCharge.DiscountEntry.Qa(_, datasets) =>
+            datasets.toList.traverse_(session.execute(StoreQaDiscountDataset)(id, _))
+          case _                                        =>
+            Applicative[F].unit
+        }
+      }
+
+      def addCorrection(
+        visitId: Visit.Id,
+        input:   TimeChargeCorrectionInput
+      )(using Transaction[F]): F[Long] =
         for {
-          o <- session.unique(SelectObservationId)(visitId)
-          c <- session.unique(SelectObserveClass)(o)
-          s <- session
-                 .stream(SelectEvents)((c, visitId), 1024)
-                 .through(TimeAccountingState.eventStreamPipe(c.chargeClass, visitId))
-                 .compile
-                 .onlyOrError
-        } yield s
+          id <- session.unique(StoreCorrection)(visitId, input, user.id)
+          _  <- session.execute(PerformCorrection)(visitId, input).void
+        } yield id
+
+      def selectObservation(
+        observationId: Observation.Id
+      )(using Transaction[F]): F[CategorizedTime] =
+        session
+          .unique(SelectObservation)(observationId)
+          .recoverWith { case SqlState.DatetimeFieldOverflow(_) =>
+            TimeAccounting.CategorizedTimeMax.pure[F]
```

Comment:
This seems sketchy in that if any of non-charged, partner or program sums overflows they are all set to the maximum time span.  I couldn't figure away around it though other than performing the query individually for each category (i.e., three times instead of once).  This is a case that in reality doesn't occur except deliberately or perhaps via a scripting error by an API user or maybe a bug in the client.